src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellmanAndroid.Derive.cs

```diff
@@ -0,0 +1,196 @@
+            private byte[]? DeriveSecretAgreement(ECDiffieHellmanPublicKey otherPartyPublicKey, IncrementalHash? hasher)
+            {
+                Debug.Assert(otherPartyPublicKey != null);
+
+                // Ensure that this ECDH object contains a private key by attempting a parameter export
+                // which will throw an OpenSslCryptoException if no private key is available
+                ECParameters thisKeyExplicit = ExportExplicitParameters(true);
+                bool thisIsNamed = Interop.AndroidCrypto.EcKeyHasCurveName(_key.Value);
+                ECDiffieHellmanAndroidPublicKey? otherKey = otherPartyPublicKey as ECDiffieHellmanAndroidPublicKey;
+                bool disposeOtherKey = false;
+
+                if (otherKey == null)
+                {
+                    disposeOtherKey = true;
+
+                    ECParameters otherParameters =
+                        thisIsNamed
+                            ? otherPartyPublicKey.ExportParameters()
+                            : otherPartyPublicKey.ExportExplicitParameters();
+
+                    otherKey = new ECDiffieHellmanAndroidPublicKey(otherParameters);
+                }
+
+                bool otherIsNamed = otherKey.HasCurveName;
+
+                SafeEcKeyHandle? ourKey = null;
+                SafeEcKeyHandle? theirKey = null;
+                byte[]? rented = null;
+                int secretLength = KeySize;
+
+                try
+                {
+                    if (otherKey.KeySize != KeySize)
+                    {
+                        throw new ArgumentException(SR.Cryptography_ArgECDHKeySizeMismatch, nameof(otherPartyPublicKey));
+                    }
+
+                    if (otherIsNamed == thisIsNamed)
+                    {
+                        ourKey = _key.UpRefKeyHandle();
+                        theirKey = otherKey.DuplicateKeyHandle();
+                    }
+                    else if (otherIsNamed)
+                    {
+                        ourKey = _key.UpRefKeyHandle();
+
+                        using (ECAndroid tmp = new ECAndroid(otherKey.ExportExplicitParameters()))
+                        {
+                            theirKey = tmp.UpRefKeyHandle();
+                        }
+                    }
+                    else
+                    {
+                        using (ECAndroid tmp = new ECAndroid(thisKeyExplicit))
+                        {
+                            ourKey = tmp.UpRefKeyHandle();
+                        }
+
+                        theirKey = otherKey.DuplicateKeyHandle();
+                    }
+
+                    // Indicate that secret can hold stackallocs from nested scopes
+                    Span<byte> secret = stackalloc byte[0];
+
+                    // Arbitrary limit. But it covers secp521r1, which is the biggest common case.
+                    const int StackAllocMax = 66;
+
+                    if (secretLength > StackAllocMax)
```

Comment:
Wasn't secretLength set to KeySize, so it's in bits whereas the StackAllocMax is in bytes?

Reply:
Yes. This looks to also be a bug in the OpenSSL implementation. I'll fix both.

Reply:
Actually nevermind the OpenSSL impl asks OpenSSL for the size.